### PR TITLE
Build local PQ in each IVF partition

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/src/index/vector/pq.rs [hunk 1]
```diff
@@ -363,7 +372,8 @@ impl ProductQuantizer {
                 flatten_array,
                 sub_vector_dimension,
                 num_centroids,
-                25,
+                50,
```
Comment (changhiskhan) [inline rust/src/index/vector/pq.rs:375]: what's 50 and 3 ?

File: rust/src/utils/kmeans.rs [hunk 2]
```diff
@@ -172,6 +172,30 @@ struct KMeanMembership {
...
+    // Used to split the centroid of largest cluster to two centroids.
+    // See Faiss
+    const EPS: f32 = 1.0 / 1024.0;
+
+    let mut histogram = histogram;
+    let mut centroids = centroids;
+    // Split the biggest clusters
+    for i in 0..histogram.len() {
+        if histogram[i] == 0 {
+            // Find the largest cluster to split into halfs.
```
Comment (changhiskhan) [inline rust/src/utils/kmeans.rs:187]: FAISS does this probabilistically, wonder why?

File: rust/src/utils/kmeans.rs [hunk 3]
```diff
@@ -172,6 +172,30 @@ struct KMeanMembership {
...
+fn split_clusters(centroids: Vec<Float32Array>, histogram: Vec<i32>) -> Result<Vec<Float32Array>> {
+    // Used to split the centroid of largest cluster to two centroids.
+    // See Faiss
+    const EPS: f32 = 1.0 / 1024.0;
+
+    let mut histogram = histogram;
+    let mut centroids = centroids;
+    // Split the biggest clusters
+    for i in 0..histogram.len() {
+        if histogram[i] == 0 {
```
Comment (changhiskhan) [inline rust/src/utils/kmeans.rs:186]: curious if any of the clusters in our test data actually gets split

File: rust/src/utils/kmeans.rs [hunk 4]
```diff
@@ -334,6 +378,33 @@ impl KMeans {
         best_kmeans
     }
 
+    /// Create a kmeans model for a dataset that has less than `k` vectors.
+    ///
+    /// It assigns each vector as a centroid, and make the rest of the centroids
+    /// to be `f32::MAX`.
+    fn new_partial_kmeans(
```
Comment (changhiskhan) [inline rust/src/utils/kmeans.rs:385]: if we have 256 pq centroids, when would this be triggered in the test data we have?

File: rust/src/utils/kmeans.rs (hunk 3 above)
Comment (eddyxu) [reply rust/src/utils/kmeans.rs:186]: For the dataset i was running on, it had 1 split on global IVF kmeans, about 10-20 splits at each local pq training.

**Pull request closed without merging**